### ISSUE TITLE
Modernize the geotargeting plugin to work with MaxMind's GeoIP2

### ISSUE THREAD

**Issue by erikgeurts** [revive-adserver/revive-adserver#315]
MaxMind is moving away from the GeoIP Lite (now called Legacy) data files and towards their new GeoIP2 file format. It might be advantageous to modernize the geotargeting plugin, so that it is able to use this new file format.

We are looking for community members who are interested in helping with this, in the form of supporting or perhaps even sponsoring the efforts to achieve this.

**Comment by Rech25** [comment]
Hi,
What is the status of GeoIP2 support ?

**Comment by andrewatfornax** [comment]
Hi @Rech25 - it's something that still needs to be done! We've done some work on it, but one of the key issues is figuring out how to migrate people's targeting data on upgrade. That's been the blocking issue, and we really just need to be able to spend more time on the issue to make progress.

**Comment by utrenkner** [comment]
Hi, we have been using OpenX and Revive for years and are now interested in geo targetting. Unfortunately, it seems the built-in solution will be dead very soon, when MadMax stops updating its GeoLite Legacy database ([in 3 months from now](http://dev.maxmind.com/geoip/legacy/geolite/)).

Is there an alternative solution, we could use?

**Comment by rhel** [comment]
> We've done some work on it, but one of the key issues is figuring out how to migrate people's targeting data on upgrade.

I guess there is no problem. You can include two plugins at the same time: geo legacy and geoip2. Each  user can select which one to use.

**Comment by rhel** [comment]
Another solution you can enable support of custom geo headers. So user is able to enable something like this https://github.com/leev/ngx_http_geoip2_module in front of revive server.

**Comment by erikgeurts** [comment]
Hi @rhel, thanks for your notes. We have a meeting scheduled with the core Revive Adserver team next week, mostly to discuss an upcoming bug fix release, but also about our longer term roadmap. This issue is also on the agenda.

Our current thinking is along the lines of your first suggestion, since it turns out to be very hard if not impossible to do a full migration of existing geotargeting delivery rules to the new version.

An important consideration is to ensure that users will have the ability to modify their delivery rules from the 'old' to the 'new' geotargeting functionality. That's why the concept of retaining the existing functionality as legacy is useful, despite the fact that in a few months the GeoLite and GeoIP .dat files will no longer be updated by MaxMind.

**Comment by erikgeurts** [comment]
The Revive Adserver project team had a project meeting today. It was decided to create a separate [release v4.2.0](https://github.com/revive-adserver/revive-adserver/milestone/22) specifically for this issue. Once we've completed the upcoming [v4.1.2 bug fix release](https://github.com/revive-adserver/revive-adserver/milestone/20), we'll start preparing the v4.2.0 release.

**Comment by ekubischta** [comment]
Some comments from an old user.  We currently run an old version that we forked off of OpenX 2.8.x

Hoepfully someday, we can reconcile our fork with Revive.  

One enhancement we made that maybe you would want to build support for, is passing Geo data over the request to the adserver.  Which overrides whatever the GeoIP lookup says.

GeoIP for cellphones is wrong...alot.   And we serve ads to our native app, which supports geolocation through GPS.  We have more accurate data than GeoIP, so, we send it in the ad call.

Might be a good addition to our update.

Eric
